src/systems/filecoin_markets/storage_market/storage_market_actor.go

```diff
@@ -242,74 +235,299 @@ func (a *StorageMarketActorCode_I) PublishStorageDeals(rt Runtime, newStorageDea
+	for _, dealID := range dealIDs {
+
+		newExpiredDeal, found := st.ActiveDeals()[dealID]
+
+		if !found {
+			// rt.Abort("sma._expireStorageDeals: dealID not found")
+			return
+		}
+
+		duration := lastChallengeEndEpoch - newExpiredDeal.LastPaymentEpoch()
+		if duration > 0 {
+			fee := actor.TokenAmount(uint64(duration) * uint64(newExpiredDeal.Deal().Proposal().StoragePricePerEpoch()))
+
+			if fee > newExpiredDeal.LockedStorageFee() {
+				// rt.Abort("sma._expireStorageDeals: cannot unlock more than already locked.")
+				return
+			}
+
+			newExpiredDeal.Impl().LastPaymentEpoch_ = lastChallengeEndEpoch
+			newExpiredDeal.Impl().UnlockStorageFee(fee)
+		}
+
+		delete(st.ActiveDeals(), dealID)
+
+		// should we check if LockedStorageFee is now 0?
+		// credit UnlockedStorageFee to miner
+		st._unlockBalance(rt, newExpiredDeal.Deal().Proposal().Provider(), newExpiredDeal.UnlockedStorageFee())
+
+		// return storage deal collaterals to both miners and client
+		st._unlockBalance(rt, newExpiredDeal.Deal().Proposal().Provider(), newExpiredDeal.ProviderCollateralRemaining())
+		st._unlockBalance(rt, newExpiredDeal.Deal().Proposal().Client(), newExpiredDeal.Deal().Proposal().TotalClientCollateral())
+
+	}
+}
+
+func (st *StorageMarketActorState_I) _creditStorageDeals(rt Runtime, dealIDs []deal.DealID, lastChallengeEndEpoch block.ChainEpoch) {
+
+	for _, dealID := range dealIDs {
+		activeDeal, found := st.ActiveDeals()[dealID]
+
+		if !found {
+			rt.Abort("sma._creditStorageDeals: dealID not found.")
+		}
+
+		duration := lastChallengeEndEpoch - activeDeal.LastPaymentEpoch()
+
+		if duration <= 0 {
+			// return for this particular deal instead of aborting the whole operation
+			// rt.Abort("sma._creditStorageDeals: no new payment to be credited.")
+			return
+		}
+
+		fee := actor.TokenAmount(uint64(duration) * uint64(activeDeal.Deal().Proposal().StoragePricePerEpoch()))
+
+		if fee > activeDeal.LockedStorageFee() {
+			// rt.Abort("sma._creditStorageDeals: cannot unlock more than already locked.")
+			return
+		}
+
+		activeDeal.Impl().LastPaymentEpoch_ = lastChallengeEndEpoch
+
+		// potentially unnecessary as we can unlock funds for provider directly
+		// unless the protocol plans on refunding client for MaxFaultCount consecutive fails
+		activeDeal.Impl().UnlockStorageFee(fee)
+
+		// TODO: align on provider storage fee withdrawal
+	}
+
+}
+
+func (st *StorageMarketActorState_I) _slashDeclaredFaults(rt Runtime, dealIDs []deal.DealID) {
+
+	for _, dealID := range dealIDs {
+
+		deal, found := st.ActiveDeals()[dealID]
+
+		if !found {
+			// move on to the next deal and keep slashing
+			return
+		}
+
+		// TODO: the exact slash amount is up for change
+		// TODO: check if provider runs out of collateral to slash here
+		deal.Impl().ProviderCollateralRemaining_ -= deal.Deal().Proposal().ProviderCollateralPerEpoch()
```

Comment:
- ideally no direct money arithmetic. Only `transfer`, `lock`, `unlock` functions that make symmetric transfers in exactly the same spot.
- sounds like this is not a transfer/lock/unlock, but merely marking collateral as spent/slashed.
- here, i would declare a function on `StorageMarketActor` 

```go
func (st *StorageMarketActorState_I) _slashDealCollateral(d *ActiveDeal) {
    st.CollateralSlashed += amt
    d.ProviderCollateralRemaining_ -= amt
}
``` 
- the money needs to be tallied up somewhere (added to some running count), exactly symmetrically.
- for easier correctness, accounting money should always be symmetric. that way everything is consistent, money numbers only "move", and are never created or destroyed. (even secondary accounts/tallies of money should be symmetric)